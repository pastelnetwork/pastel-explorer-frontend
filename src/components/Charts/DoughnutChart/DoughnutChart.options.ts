import themeVariant from '@theme/variants';

export const chartColors = [
  themeVariant.palette.primary.main,
  themeVariant.palette.secondary.main,
  themeVariant.custom.green.light,
  themeVariant.custom.red.light,
  themeVariant.custom.orange.light,
  themeVariant.custom.green.dark,
  themeVariant.custom.red.dark,
  themeVariant.custom.orange.dark,
];
