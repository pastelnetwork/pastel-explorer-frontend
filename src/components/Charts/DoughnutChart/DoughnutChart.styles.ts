import styled from 'styled-components/macro';
import { Link as RouterLink } from 'react-router-dom';

import themeVariant from '@theme/variants';

import {
  Card as MuiCard,
  CardContent as MuiCardContent,
  TableCell as MuiTableCell,
  TableRow as MuiTableRow,
  Grid,
} from '@material-ui/core';

export const Card = styled(MuiCard)`
  background: transparent;
  box-shadow: none;

  h4:not(.MuiTypography-h4) {
    margin: 0;
    padding-top: 18px;
    padding-bottom: 18px;
    padding-left: 16px;
    background: ${props => props.theme.card.titleColor};
  }
`;

export const CardContent = styled(MuiCardContent)`
  @media (min-width: 960px) {
    max-height: 356px;
  }

  &:last-child {
    padding-top: 0;
    padding-bottom: ${props => props.theme.spacing(2)}px;
  }

  ${props => props.theme.breakpoints.down('sm')} {
    & > .MuiGrid-root {
      align-items: center;
    }
  }

  ${props => props.theme.breakpoints.down(600)} {
    &:last-child {
      padding-top: 12px;
    }
  }
`;

export const ChartWrapper = styled(Grid)`
  height: 160px;
  position: relative;
  padding: 20px;

  @media (min-width: 960px) {
    padding: 10px 20px;
    height: 273px;
  }

  @media (min-width: 1280px) {
    padding: 20px;
    height: 332px;
  }

  .chartjs-render-monitor {
    position: relative;
  }

  canvas {
    position: relative;
    z-index: 1;
    }
  }
`;

export const DoughnutInner = styled.div`
  position: absolute;
  top: 50%;
  left: 50%;
  margin-top: -22px;
  text-align: center;
  z-index: 0;
  width: auto;
  transform: translateX(-50%);
`;

export const TableRow = styled(MuiTableRow)`
  height: 42px;

  &:last-child th,
  &:last-child td {
    border-bottom: 0;
  }
`;

export const TableCell = styled(MuiTableCell)`
  padding-top: 0;
  padding-bottom: 0;
`;

export const GreenText = styled.span`
  color: ${() => themeVariant.custom.green.main};
`;

export const RedText = styled.span`
  color: ${() => themeVariant.custom.red.main};
`;

export const StakingWrapper = styled.div`
  padding: 20px;
  text-align: center;
  font-size: 60px;
  font-weight: 600;
  line-height: 1;
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;

export const StakingTitle = styled.div`
  margin-top: 15px;
  font-size: 20px;
  font-weight: 400;
  line-height: 1;
`;

export const Link = styled(RouterLink)`
  text-decoration: none;
  color: ${props => props.theme.palette.text.primary};
`;
