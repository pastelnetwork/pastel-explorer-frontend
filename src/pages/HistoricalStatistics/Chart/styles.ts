import { makeStyles } from '@material-ui/core/styles';
import { TAppTheme } from '@theme/index';

export const eChartLineStyles = makeStyles((theme: TAppTheme) => ({
  container: {
    margin: '0 auto',
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    height: '100%',
    width: '100%',
    padding: '0 20px',
    textAlign: 'center',
    boxSizing: 'border-box',
    position: 'relative',
    color: 'rgb(171, 170, 193)',
    [theme.breakpoints.down('sm')]: {
      padding: 0,
    },
  },
  lineChartWrap: {
    width: '100%',
    height: 'auto',
    marginBottom: 20,
  },
  lineChartContainer: {
    display: 'block',
    width: '100%',
    height: '100%',
    backgroundColor: '#212121',
  },
  lineChartTitle: {
    width: '100%',
    textAlign: 'left',
    fontSize: '30px',
    [theme.breakpoints.down('sm')]: {
      fontSize: 15,
    },
  },
  lineChartHeader: {
    display: 'flex',
    width: '100%',
    justifyContent: 'space-between',
    paddingTop: '30px',
    paddingBottom: '10px',
    [theme.breakpoints.down('sm')]: {
      paddingTop: 10,
      paddingLeft: 10,
      justifyContent: 'flex-start',
      flexDirection: 'column',
    },
  },
  uploadButton: {
    display: 'flex',
    alignItems: 'center',
    backgroundColor: '#6151d0',
    color: 'white',
    textAlign: 'center',
    fontWeight: 'bold',
    fontSize: '14px',
    padding: '8px 16px',
    borderRadius: '4px',
    marginLeft: '8px',
    marginRight: '8px',
    '&:hover': {
      cursor: 'pointer',
      opacity: 0.8,
    },
  },
  reactECharts: {
    width: '100%',
    height: 'calc(100vh - 450px) !important',
  },
  granularitySelect: {
    display: 'flex',
    width: '100%',
    maxWidth: 200,
    justifyContent: 'space-around',
    alignItems: 'center',
    marginTop: 15,
    marginRight: 50,
  },
  periodSelect: {
    display: 'flex',
    minWidth: '200px',
    // justifyContent: 'space-around',
    alignItems: 'center',
    marginTop: '15px',
    [theme.breakpoints.down('sm')]: {
      marginTop: 5,
    },
  },
  lineChartDownloadButtonBar: {
    display: 'flex',
    width: '100%',
    justifyContent: 'flex-end',
    height: '30px',
    [theme.breakpoints.down('sm')]: {
      marginTop: 10,
      justifyContent: 'flex-start',
    },
  },
  lineChartFooter: {
    display: 'flex',
    width: '100%',
    justifyContent: 'space-around',
    alignItems: 'center',
    paddingBottom: 30,
    [theme.breakpoints.down('sm')]: {
      flexWrap: 'wrap',
      justifyContent: 'flex-start',
      paddingBottom: 10,
    },
  },
  lineChartThemeSelect: {
    display: 'flex',
    minWidth: '150px',
    justifyContent: 'space-around',
    alignItems: 'center',
  },
  themeSelectButton: {
    width: '30px',
    height: '30px',
    borderRadius: '5px',
  },
  filterButton: {
    backgroundColor: '#eee',
    color: '#212124',
    textAlign: 'center',
    fontWeight: 'bold',
    fontSize: '12px',
    padding: '4px 8px',
    borderRadius: '3px',
    borderWidth: '1px',
    borderColor: '#212124',
    marginLeft: '4px',
    marginRight: '4px',
    '&:hover': {
      cursor: 'pointer',
      opacity: '0.8',
    },
  },
  activeButton: {
    backgroundColor: '#c3921f',
    borderColor: '#c3921f',
    color: '#fff',
  },
  activeThemeButton: {
    borderColor: '#008000',
    borderWidth: '5px',
  },
}));
