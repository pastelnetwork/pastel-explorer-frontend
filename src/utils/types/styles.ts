/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/ban-types */
import { SpacingProps } from '@material-ui/system';
import { Theme } from '@material-ui/core';

export type GlobalStyleProps = {
  theme: Theme & { palette: any };
};

export interface MuiButtonSpacingType extends SpacingProps {
  component?: React.PropsWithoutRef<{}>;
  to?: string;
}
