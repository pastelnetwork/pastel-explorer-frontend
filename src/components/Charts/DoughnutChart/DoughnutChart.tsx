import React from 'react';
import ReactECharts from 'echarts-for-react';
import { Typography, Grid } from '@material-ui/core';

import { translate } from '@utils/helpers/i18n';
import { TChartParams } from '@utils/types/IStatistics';
import themeVariant from '@theme/variants';

import { chartColors } from './DoughnutChart.options';
import * as Styles from './DoughnutChart.styles';

type TDatasetsProps = {
  data: Array<number>;
  backgroundColor: string[];
  borderWidth: number;
  borderColor: string;
  barThickness: number;
};

type TChartDataProps = {
  labels: Array<string>;
  datasets: TDatasetsProps[];
};

interface DoughnutChartProps {
  data: TChartDataProps | null;
  title?: string;
  innerTitle?: string;
  innerSubtitle?: string | number;
  totalSuperNodes?: number;
  link?: string;
}

const DoughnutChart: React.FC<DoughnutChartProps> = ({
  data,
  title,
  innerTitle,
  innerSubtitle,
  totalSuperNodes,
  link = '',
}) => {
  const transferChartData = () => {
    if (!data) {
      return [];
    }

    return data?.labels?.map((label, index) => ({
      value: data.datasets[0].data[index],
      name: label,
    }));
  };
  const options = {
    color: chartColors,
    tooltip: {
      trigger: 'item',
      formatter(params: TChartParams) {
        return `${params.marker} ${params.name}: ${params.value}(${params.percent}%)`;
      },
      backgroundColor: themeVariant.palette.text.primary,
      textStyle: {
        color: themeVariant.custom.white,
        fontSize: 12,
      },
    },
    series: [
      {
        type: 'pie',
        radius: ['70%', '90%'],
        avoidLabelOverlap: true,
        label: {
          show: false,
          position: 'center',
        },
        itemStyle: {
          borderColor: '#fff',
          borderWidth: 1,
        },
        emphasis: {
          scaleSize: 0,
          label: {
            show: false,
          },
        },
        labelLine: {
          show: true,
        },
        data: transferChartData(),
      },
    ],
  };

  const renderChart = () => {
    return (
      <Grid container justify="center" alignItems="flex-start">
        <Styles.ChartWrapper item xs={12} sm={6} md={5}>
          <Styles.DoughnutInner>
            {innerTitle && <Typography variant="h4">{innerTitle}</Typography>}
            {innerSubtitle && <Typography variant="caption">{innerSubtitle}</Typography>}
          </Styles.DoughnutInner>
          {data && <ReactECharts notMerge={false} lazyUpdate option={options} />}
        </Styles.ChartWrapper>
        <Styles.ChartWrapper item xs={12} sm={6} md={7}>
          <Styles.StakingWrapper>
            {totalSuperNodes ? `${((51.84 / totalSuperNodes) * 100).toFixed(2)}%` : '--'}
            <Styles.StakingTitle>
              {translate('components.charts.doughnutChart.stakingAPR')}
            </Styles.StakingTitle>
          </Styles.StakingWrapper>
        </Styles.ChartWrapper>
      </Grid>
    );
  };
  return (
    <Styles.Card>
      {title && <h4>{title}</h4>}
      <Styles.CardContent>
        {link ? <Styles.Link to={link}>{renderChart()}</Styles.Link> : renderChart()}
      </Styles.CardContent>
    </Styles.Card>
  );
};

export default DoughnutChart;
