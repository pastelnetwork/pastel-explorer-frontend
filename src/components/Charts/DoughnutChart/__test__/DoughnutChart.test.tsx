import { shallow } from 'enzyme';
import { Typography, Grid } from '@mui/material';
import ReactECharts from 'echarts-for-react';

import 'jest-styled-components';

// import i18next from '../../../../utils/helpers/i18n';
import DoughnutChart from '../DoughnutChart';
import * as Styles from '../DoughnutChart.styles';

jest.mock('i18next-http-backend');
jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () =>
          new Promise(() => {
            // noop
          }),
      },
    };
  },
  initReactI18next: {
    type: '3rdParty',
    init: () => {
      // noop
    },
  },
}));
jest.mock('i18next', () => ({
  t: (str: string) => str,
}));
// i18next.t = jest.fn().mockImplementation((...arg) => {
//   return arg[0];
// });

describe('components/Charts/DoughnutChart', () => {
  const data = {
    labels: ['United States'],
    datasets: [
      {
        data: [27],
        backgroundColor: [
          '#279989',
          '#41B6E6',
          '#a5d6a7',
          '#ef9a9a',
          '#ffcc80',
          '#2e7d32',
          '#c62828',
          '#ef6c00',
        ],
        borderWidth: 1,
        borderColor: '#FFF',
        barThickness: 40,
      },
    ],
  };
  const wrapper = shallow(
    <DoughnutChart
      data={data}
      title="Supernode Statistics"
      innerTitle="Total"
      totalSuperNodes={27}
    />,
  );

  test('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should render <ReactECharts>', () => {
    expect(wrapper.find(ReactECharts).length).toBeGreaterThanOrEqual(1);
  });

  test('should render <Grid>', () => {
    expect(wrapper.find(Grid).length).toBeGreaterThanOrEqual(1);
  });

  test('should render <MuiAlert>', () => {
    expect(wrapper.find(Typography).length).toBeGreaterThanOrEqual(1);
  });

  test('should render <Styles.ChartWrapper>', () => {
    expect(wrapper.find(Styles.ChartWrapper).length).toBeGreaterThanOrEqual(1);
  });

  test('should render <Styles.DoughnutInner>', () => {
    expect(wrapper.find(Styles.DoughnutInner).length).toBeGreaterThanOrEqual(1);
  });

  test('should render <Styles.StakingWrapper>', () => {
    expect(wrapper.find(Styles.StakingWrapper).length).toBeGreaterThanOrEqual(1);
  });

  test('should render <Styles.StakingTitle>', () => {
    expect(wrapper.find(Styles.StakingTitle).length).toBeGreaterThanOrEqual(1);
  });

  test('should render <Styles.CardContent>', () => {
    expect(wrapper.find(Styles.CardContent).length).toBeGreaterThanOrEqual(1);
  });
});
