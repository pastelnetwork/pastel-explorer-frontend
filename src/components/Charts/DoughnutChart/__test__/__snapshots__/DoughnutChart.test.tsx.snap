// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Charts/DoughnutChart renders correctly 1`] = `ShallowWrapper {}`;
